libgeopmd/src/RuntimeStats.cpp

```diff
@@ -0,0 +1,181 @@
+{
+
+    RuntimeStats::RuntimeStats(const std::vector<std::string> &metric_names)
+        : m_metric_names(metric_names)
+        , m_moments(m_metric_names.size())
+    {
+        reset();
+    }
+
+    int RuntimeStats::num_metric(void) const
+    {
+        return m_metric_names.size();
+    }
+
+    void RuntimeStats::check_index(int metric_idx, const std::string &func, int line) const
+    {
+        if (metric_idx < 0 || (size_t)metric_idx >= m_metric_names.size()) {
+            throw Exception("RuntimeStats::" + func  + "(): metric_idx out of range: " + std::to_string(metric_idx),
+                            GEOPM_ERROR_INVALID, __FILE__, line);
+        }
+    }
+
+    std::string RuntimeStats::metric_name(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        return m_metric_names[metric_idx];
+    }
+
+    uint64_t RuntimeStats::count(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        return m_moments[metric_idx].count;
+    }
+
+    double RuntimeStats::first(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        double result = NAN;
+        if (m_moments[metric_idx].count != 0) {
+            result = m_moments[metric_idx].first;
+        }
+        return result;
+    }
+
+    double RuntimeStats::last(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        double result = NAN;
+        if (m_moments[metric_idx].count != 0) {
+            result = m_moments[metric_idx].last;
+        }
+        return result;
+    }
+
+    double RuntimeStats::min(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        double result = NAN;
+        if (m_moments[metric_idx].count != 0) {
+            result = m_moments[metric_idx].min;
+        }
+        return result;
+    }
+
+    double RuntimeStats::max(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        double result = NAN;
+        if (m_moments[metric_idx].count != 0) {
+            result = m_moments[metric_idx].max;
+        }
+        return result;
+    }
+
+    double RuntimeStats::mean(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        double result = NAN;
+        if (m_moments[metric_idx].count != 0) {
+            result = m_moments[metric_idx].m_1 /
+                     m_moments[metric_idx].count;
+        }
+        return result;
+    }
+
+    double RuntimeStats::std(int metric_idx) const
+    {
+        check_index(metric_idx, __func__, __LINE__);
+        double result = NAN;
+        if (m_moments[metric_idx].count > 1) {
+            result = std::sqrt(
+                         (m_moments[metric_idx].m_2 -
+                          m_moments[metric_idx].m_1 *
+                          m_moments[metric_idx].m_1 / m_moments[metric_idx].count) /
+                         (m_moments[metric_idx].count - 1));
```

Comment:
I was expecting the second term inside the sqrt to be minus mean-squared:
```
 - m_moments[metric_idx].m_1 * m_moments[metric_idx].m_1
   / (m_moments[metric_idx].count * m_moments[metric_idx].count)
```

But I'm not sure if I'm interpreting m_1 and m_2 correctly. Need to return after reviewing the rest of this class

Reply:
Coming back here, I think we have
```
std = sqrt(N*moment_2 - N*moment_1 * N*moment_1 / N) / (N-1)
    = sqrt(N*moment_2 - N*moment_1^2) / (N-1)
    = sqrt(N) * sqrt(moment_2 - moment_1^2) / (N-1)
```

But I think we're aiming for this?
```
std = sqrt(moment_2 - moment_1^2)
```

I think I'm a bit lost. Can you point me to which formula I should be using?

Reply:
This issue with your formula is the bias:

https://en.wikipedia.org/wiki/Unbiased_estimation_of_standard_deviation

I can't find my references.  It can be derived, but you use up a degree of freedom by using m_1 and m_2. This is where the n-1 is coming from.

Reply:
The n-1 is used in situations where you wish to extrapolate, i.e. you are estimating the stdev of a distribution based on a set of samples. If you wish to get the stdev of a set of samples without wishing to extrapolate, you can use n. It's not clear to me which is more appropriate here. Is the purpose of the reporter to report statistics of the samples that we've taken, or to understand the characteristics of the hw/wkld?

Reply:
What about the extra `sqrt(n)`?

Reply:
I think the problem is that I misaligned parentheses when translating from the code to my first line above. Instead I think it translates to:

```
std = sqrt( (N*moment_2 - N*moment_1 * N*moment_1 / N) / (N-1) ) <-- I accidentally moved N-1 outside sqrt earlier
    = sqrt( (N*moment_2 - N*moment_1^2 ) / (N-1) )
```

Which works out for sample standard deviation